### ISSUE TITLE
Watcom math functions

### ISSUE THREAD

**Issue by slouken** [libsdl-org/SDL#5667]
Let's fix the math functions in SDL so they work with Watcom. Based on previous commits, it looks like they are available and work, we just need SDL to call them instead of using the uClibc versions, is that right? Is the issue with the SDL implementations a Watcom code generation bug?

I will be going through and removing the Watcom math ifdefs in various SDL libraries once we sort this out.

**Comment by sezero** [comment]
As I responded at the SDL_image commit note, all of C89 math functions
are available in watcom, and the SDL_xxx math functions do already call
them.

The issue with satellite projects' dlls calling SDL_xxx math functions:
I _think_ the issue is about SDL being linked statically to OpenWatcom's
runtime: SDL itself is OK because SDL_ceil calls ceil and ceil is linked
in statically, along with whatever watcom FPU CRT stuff.  But, SDL_image
doesn't know about SDL_ceil and any FPU magic watcom would have possibly
played.

If you know a fix, it'd be welcome.

**Comment by slouken** [comment]
What happens if SDL_image calls SDL_ceilf()? In theory the call should be routed into SDL2.dll and in turn call into Watcom's math function, with whatever magic is happening to make math calls in SDL, right?

**Comment by sezero** [comment]
I can test shortly in a VM, but I remember having discovered the issue in SDL_image::svg
where showimage displayed something broken (was an empty image? can't remember.)

**Comment by slouken** [comment]
Maybe it was the float word order issue that was fixed in https://github.com/libsdl-org/SDL/commit/b398a847be0814ea36fe72da4f6ecb2e6d49a86b?

**Comment by sezero** [comment]
> Maybe it was the float word order issue that was fixed in [b398a84](https://github.com/libsdl-org/SDL/commit/b398a847be0814ea36fe72da4f6ecb2e6d49a86b)?

Well, it's x86 where both cpu and fpu are both little endian

**Comment by sezero** [comment]
And _maybe_ not all but _some_ math functions are problematic,
(I can't find the correspondence about the original issue, was it here
or was it with @icculus? whatever), so I play safe for the time being.

**Comment by slouken** [comment]
Let's back out the change and root cause the issue. What you've done right now means that all code anywhere using SDL math functions needs to add an #ifdef for WATCOM, and that's not feasible. All applications calling SDL math functions will fail as it is currently.

**Comment by sezero** [comment]
> What happens if SDL_image calls SDL_ceilf()?

Just tried in a VM with watcom ifdef disabled, showimage attempting load
a svg file exited with `Out of memory`, whereas the dll with watcom ifdefs
in place displayed the image fine.

**Comment by sezero** [comment]
> Let's back out the change 

I recommend against it, not until a solution is found.

**Comment by slouken** [comment]
We should not disable the math functions on Watcom until we understand why they're not working. As I said, we don't know what's happening, and it's not feasible to ask everyone using Watcom to add ifdefs to their code. Let's find out the root cause and solution. Please don't put Watcom hacks back into the SDL libraries.

**Comment by sezero** [comment]
Calling convention _can_ be an issue:
```c
#include <math.h>
float _CALLCONV_ ceilf(float x) {
  return (float) ceil((double)x);
}
```

The generated asm difference between _CALLCONV_ being the standard `__watcall`
(default of watcom) and `_System` (default of SDL) is this:
```diff
--- watcall.asm
+++ syscall.asm
@@ -11,14 +11,12 @@
 ceilf:
-	sub		esp,0cH
-	fld		dword ptr 10H[esp]
+	sub		esp,8
+	fld		dword ptr 0cH[esp]
 	sub		esp,8
 	fstp		qword ptr [esp]
 	call		near ptr FLAT:ceil
 	mov		dword ptr 8[esp],eax
 	mov		dword ptr 0cH[esp],edx
 	fld		qword ptr 8[esp]
 	add		esp,8
-	fstp		dword ptr 8[esp]
-	mov		eax,dword ptr 8[esp]
-	add		esp,0cH
+	add		esp,8
 	ret		
```

**Comment by sezero** [comment]
Something with dynapi... If I do the following to disable dynapi (the second part
in order to workaround calling convention differences), then `showimage logo.svg`
works:
```diff
diff --git a/src/dynapi/SDL_dynapi.h b/src/dynapi/SDL_dynapi.h
index e37444f..9b49077 100644
--- a/src/dynapi/SDL_dynapi.h
+++ b/src/dynapi/SDL_dynapi.h
@@ -61,6 +61,8 @@
 #define SDL_DYNAMIC_API 0  /* vitasdk doesn't support dynamic linking */
 #elif defined(__NGAGE__)
 #define SDL_DYNAMIC_API 0  /* The N-Gage doesn't support dynamic linking either */
+#elif defined(__OS2__)
+#define SDL_DYNAMIC_API 0
 #elif defined(DYNAPI_NEEDS_DLOPEN) && !defined(HAVE_DLOPEN)
 #define SDL_DYNAMIC_API 0  /* we need dlopen(), but don't have it.... */
 #endif
diff --git a/src/stdlib/SDL_malloc.c b/src/stdlib/SDL_malloc.c
index b5dd641..8a64e48 100644
--- a/src/stdlib/SDL_malloc.c
+++ b/src/stdlib/SDL_malloc.c
@@ -5305,10 +5305,14 @@ History:
 #endif /* !HAVE_MALLOC */
 
 #ifdef HAVE_MALLOC
-#define real_malloc malloc
-#define real_calloc calloc
-#define real_realloc realloc
-#define real_free free
+static void* SDLCALL SDL2_malloc(size_t sz) { return malloc(sz); }
+static void* SDLCALL SDL2_calloc(size_t n, size_t sz) { return calloc(n, sz); }
+static void* SDLCALL SDL2_realloc(void *p, size_t sz) { return realloc(p,sz); }
+static void  SDLCALL SDL2_free(void *p) { free(p); }
+#define real_malloc SDL2_malloc
+#define real_calloc SDL2_calloc
+#define real_realloc SDL2_realloc
+#define real_free SDL2_free
 #else
 #define real_malloc dlmalloc
 #define real_calloc dlcalloc
```

Can anyone see what's wrong with dynapi being enabled for os/2?

**Comment by sezero** [comment]
@icculus, @slouken: need help with seeing what's wrong with os/2 dynapi here
(I've been staring at the same thing for an hour, can't see it..)

**Comment by sezero** [comment]
And I can happily disable dynapi for os/2 if no other solution is found 
(the SDL_malloc patch above will be needed though.)

**Comment by slouken** [comment]
I'm fine with disabling dynapi on OS/2, but I'm guessing it has something to do with the calling convention for the dispatched functions. Maybe the _REAL functions are declared with SDLCALL in one place, but not in the place they're actually implemented? This is probably worth tracking down in case it affects other platforms as well.

**Comment by sezero** [comment]
Disabling dynapi seems to fix loading and displaying svg files under a vm for me,
so I can do it now. Is the patch above OK -- the part touching SDL_malloc.c?

As for dynapi, maybe I missed some SDLCALL in the SDL_DYPAPI macros in
SDL_dynapi.c? I don't know -- can't really see it.

**Comment by slouken** [comment]
Yes, the above patch for SDL_malloc.c is fine, but why is it needed?

**Comment by sezero** [comment]
> Yes, the above patch for SDL_malloc.c is fine

OK, will apply shortly

> , but why is it needed?

Compiler didn't care before (mystery), but it does with dynapi disabled:
```
./src/stdlib/SDL_malloc.c(5328): Error! E1010: Type mismatch
./src/stdlib/SDL_malloc.c(5328): Note! I2003: source conversion type is 'void *(__watcall *)(unsigned int __p1)'
./src/stdlib/SDL_malloc.c(5328): Note! I2004: target conversion type is 'void *(_Syscall *)(unsigned int __p1)'
./src/stdlib/SDL_malloc.c(5328): Error! E1010: Type mismatch
./src/stdlib/SDL_malloc.c(5328): Note! I2003: source conversion type is 'void *(__watcall *)(unsigned int __p1,unsigned int __p2)'
./src/stdlib/SDL_malloc.c(5328): Note! I2004: target conversion type is 'void *(_Syscall *)(unsigned int __p1,unsigned int __p2)'
./src/stdlib/SDL_malloc.c(5328): Error! E1010: Type mismatch
./src/stdlib/SDL_malloc.c(5328): Note! I2003: source conversion type is 'void *(__watcall *)(void *__p1,unsigned int __p2)'
./src/stdlib/SDL_malloc.c(5328): Note! I2004: target conversion type is 'void *(_Syscall *)(void *__p1,unsigned int __p2)'
./src/stdlib/SDL_malloc.c(5328): Error! E1010: Type mismatch
./src/stdlib/SDL_malloc.c(5328): Note! I2003: source conversion type is 'void (__watcall *)(void *__p1)'
./src/stdlib/SDL_malloc.c(5328): Note! I2004: target conversion type is 'void (_Syscall *)(void *__p1)'
Error(E42): Last command making (SDL_malloc.obj) returned a bad status
```

**Comment by sezero** [comment]
Patches applied to both main and to 2.0.22 branch.

Closing this, the dynapi issue is tracked at #5669

**Comment by icculus** [comment]
> Yes, the above patch for SDL_malloc.c is fine, but why is it needed?

Because we do this:

```c
static struct
{
    SDL_malloc_func malloc_func;
    SDL_calloc_func calloc_func;
    SDL_realloc_func realloc_func;
    SDL_free_func free_func;
    SDL_atomic_t num_allocations;
} s_mem = {
    real_malloc, real_calloc, real_realloc, real_free, { 0 }
};
```

...and `SDL_malloc_func` is `SDLCALL`, which is almost universally defined as nothing (aka __cdecl), but when it's something else, like Watcom does, we can't just assign `malloc` to it without a wrapper.

Which is to say, the wrapper is not just a workaround, but the correct thing to do on all platforms...but the other branch, where we do this with dlmalloc, has the same problem; they aren't SDLCALL functions.

**Comment by sezero** [comment]
> tl;dr: the patch to SDL_malloc.c is correct, and if this was the only problem, we can turn the dynamic API on again for Watcom builds. If there are other problems, let me know and I'll fix them in #5669.

I did try with the malloc patch alone but it did not solve the issue.

**Comment by sezero** [comment]
> Which is to say, the wrapper is not just a workaround, but the correct thing to do on all platforms...but the other branch, where we do this with dlmalloc, has the same problem; they aren't SDLCALL functions.

Maybe the dlmalloc stuff should be SDLCALL unless there is compelling reason against it?

**Comment by sezero** [comment]
> ...because in src/SDL_internal.h ...
> 
> ```c
> #if SDL_DYNAMIC_API
> #include "dynapi/SDL_dynapi_overrides.h"
> /* force DECLSPEC and SDLCALL off...it's all internal symbols now.
>    These will have actual #defines during SDL_dynapi.c only */
> #define DECLSPEC
> #define SDLCALL
> #endif
> ```
> 
> ...because we force everything to use the same calling convention internally (the `_REAL` versions, which SDL can call directly from one source file to another), and only force specific calling conventions on things calling into the library (which, when using the dynamic API, is just the little entry point jump code thing).

I don't fully understand this (and the actual watcom issue is most possibly because of this):
`DECLSPEC` is OK, but why really nullify `SDLCALL` there?

**Comment by icculus** [comment]
> DECLSPEC is OK, but why really nullify SDLCALL there?

Actually, that's a good question. It maybe shouldn't be.

My thinking was probably that the compiler knows better for calls that never leave the DLL, and the SDLCALL calling conventions were just to make sure external code interacts correctly, but that could definitely be a faulty assumption.

**Comment by ccawley2011** [comment]
> 

From what I can tell, it's necessary for entries in the jump table to use `SDLCALL` conventions so that `SDL_DYNAPI_entry` is ABI compatible between different compilers, and this requires the `_REAL` functions to also use `SDLCALL` so that they can be used in the jump table. As a result, the `_REAL` functions are declared with `SDLCALL` in `SDL_dynapi.c`, but because `SDLCALL` is undefined in the other files, the actual functions use the default calling conventions instead.

**Comment by slouken** [comment]
@sezero, would this also fix the math functions on OS/2?

**Comment by sezero** [comment]
> From what I can tell, it's necessary for entries in the jump table to use `SDLCALL` conventions so that `SDL_DYNAPI_entry` is ABI compatible between different compilers, and this requires the `_REAL` functions to also use `SDLCALL` so that they can be used in the jump table. As a result, the `_REAL` functions are declared with `SDLCALL` in `SDL_dynapi.c`, but because `SDLCALL` is undefined in the other files, the actual functions use the default calling conventions instead.

Exactly: I think this is the real issue behind the broken math functions with watcom+os/2, so I think that empty SDLCALL define should go away from SDL_internal.h 

> @sezero, would this also fix the math functions on OS/2?

Commented out that SDLCALL define in SDL_internal.h, enabled dynapi: showimage.exe does properly display the 23.svg example image from nanosvg site in my VM.

**Comment by ccawley2011** [comment]
OK, I've added the fix to PR #5687, and I can confirm this works on Windows as well, both when using a DLL built with OpenWatcom, and when using one built with MSVC 2022 (including when setting the `SDL_DYNAMIC_API` environment variable).

**Comment by slouken** [comment]
Yay!!
